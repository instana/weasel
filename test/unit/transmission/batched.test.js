jest.mock('../../../lib/browser');
jest.mock('../../../lib/vars');

jest.useFakeTimers();

const firstBeacon = {
  ty: 'pl',
  page: 'landing',
  ts: 123456789,
  error: false
};

const secondBeacon = {
  ty: 'xhr',
  page: 'cart',
  ts: 123456789,
  error: true
};

describe('transmission/batched', () => {
  let browserMock;
  let varsMock;
  let sendBeacon;

  beforeEach(() => {
    browserMock = require('../../../lib/browser');
    browserMock.reset();
    varsMock = require('../../../lib/vars').default;
    sendBeacon = require('../../../lib/transmission/batched').sendBeacon;
  });

  describe('with unsupported sendBeacon API', () => {
    tests();
  });

  describe('with sendBeacon API', () => {
    beforeEach(() => {
      jest.resetModules();

      browserMock = require('../../../lib/browser');
      browserMock.reset();
      // simulate availability of the sendBeacon API
      browserMock.nav.sendBeacon = () => {};
      varsMock = require('../../../lib/vars').default;

      sendBeacon = require('../../../lib/transmission/batched').sendBeacon;
    });

    afterEach(() => {
      delete browserMock.nav.sendBeacon;
    });

    tests();
  });

  function tests() {
    describe('without batching', () => {
      it('must do nothing when there is nothing to transmit', () => {
        sendBeacon({});
        assertMatchesSnapshot();
      });

      it('must only transmit own props', () => {
        const parentObject = {};
        parentObject.shouldNotBeIncluded = 'secr3t!';
        const beacon = Object.create(parentObject);
        beacon.page = 'cart';
        sendBeacon(beacon);
        assertMatchesSnapshot();
      });

      it('must transmit larger payloads', () => {
        const beacon = {
          page: 'cart',
          id: 42,
          m_featureFlags: JSON.stringify({developerMode: false}),
          longKey: Array(2001).fill('a').join('')
        };

        sendBeacon(beacon);
        assertMatchesSnapshot();
      });

      it('must transmit multiple beacons', () => {
        sendBeacon(firstBeacon);
        sendBeacon(secondBeacon);
        assertMatchesSnapshot();
      });
    });

    describe('with batching', () => {
      beforeEach(() => {
        varsMock.beaconBatchingTime = 100;
      });

      afterEach(() => {
        varsMock.beaconBatchingTime = 0;
      });

      it('must not attempt batching when window is hidden', () => {
        try {
          setVisibilityState('hidden');
          sendBeacon(firstBeacon);
          sendBeacon(secondBeacon);
          assertMatchesSnapshot();
        } finally {
          setVisibilityState('prerender');
        }
      });

      it('must batch when window is visible', () => {
        try {
          setVisibilityState('visible');
          sendBeacon(firstBeacon);
          sendBeacon(secondBeacon);
          jest.runOnlyPendingTimers();
          assertMatchesSnapshot();
        } finally {
          setVisibilityState('prerender');
        }
      });

      it('must send early when window becomes hidden', () => {
        try {
          setVisibilityState('visible');
          sendBeacon(firstBeacon);
          sendBeacon(secondBeacon);
          assertMatchesSnapshot();
          setVisibilityState('hidden');
          assertMatchesSnapshot();
        } finally {
          setVisibilityState('prerender');
        }
      });

      it('must send early when window is unloaded', () => {
        try {
          setVisibilityState('visible');
          sendBeacon(firstBeacon);
          sendBeacon(secondBeacon);
          assertMatchesSnapshot();
          browserMock.win.dispatchEvent(new Event('beforeunload'));
          assertMatchesSnapshot();
        } finally {
          setVisibilityState('prerender');
        }
      });

      it('must have an upper limit on the number of batched beacons', () => {
        try {
          setVisibilityState('visible');
          for (let i = 0; i < 15; i++) {
            sendBeacon(firstBeacon);
          }
          for (let i = 0; i < 5; i++) {
            sendBeacon(secondBeacon);
          }
          assertMatchesSnapshot();
          browserMock.win.dispatchEvent(new Event('beforeunload'));
          assertMatchesSnapshot();
        } finally {
          setVisibilityState('prerender');
        }
      });
    });
  }

  function assertMatchesSnapshot() {
    expect({
      imageRequests: browserMock.imageRequests,
      xhrRequests: browserMock.xhrRequests,
      beaconRequests: browserMock.beaconRequests
    }).toMatchSnapshot();
  }

  function setVisibilityState(state) {
    Object.defineProperty(browserMock.doc, 'visibilityState', {
      value: state,
      configurable: true
    });

    browserMock.doc.dispatchEvent(new Event('visibilitychange'));
  }
});
