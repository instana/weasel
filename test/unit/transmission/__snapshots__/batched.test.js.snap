exports[`transmission/batched with sendBeacon API with batching must batch when window is visible 1`] = `
Object {
  "beaconRequests": Array [
    Object {
      "data": "ty	pl
page	landing
ts	123456789
error	false

ty	xhr
page	cart
ts	123456789
error	true",
      "url": "http://eum.example.com",
    },
  ],
  "imageRequests": Array [],
  "xhrRequests": Array [],
}
`;

exports[`transmission/batched with sendBeacon API with batching must not attempt batching when window is hidden 1`] = `
Object {
  "beaconRequests": Array [
    Object {
      "data": "ty	pl
page	landing
ts	123456789
error	false",
      "url": "http://eum.example.com",
    },
    Object {
      "data": "ty	xhr
page	cart
ts	123456789
error	true",
      "url": "http://eum.example.com",
    },
  ],
  "imageRequests": Array [],
  "xhrRequests": Array [],
}
`;

exports[`transmission/batched with sendBeacon API with batching must send early when window becomes hidden 1`] = `
Object {
  "beaconRequests": Array [],
  "imageRequests": Array [],
  "xhrRequests": Array [],
}
`;

exports[`transmission/batched with sendBeacon API with batching must send early when window becomes hidden 2`] = `
Object {
  "beaconRequests": Array [
    Object {
      "data": "ty	pl
page	landing
ts	123456789
error	false

ty	xhr
page	cart
ts	123456789
error	true",
      "url": "http://eum.example.com",
    },
  ],
  "imageRequests": Array [],
  "xhrRequests": Array [],
}
`;

exports[`transmission/batched with sendBeacon API without batching must do nothing when there is nothing to transmit 1`] = `
Object {
  "beaconRequests": Array [],
  "imageRequests": Array [],
  "xhrRequests": Array [],
}
`;

exports[`transmission/batched with sendBeacon API without batching must only transmit own props 1`] = `
Object {
  "beaconRequests": Array [
    Object {
      "data": "page	cart",
      "url": "http://eum.example.com",
    },
  ],
  "imageRequests": Array [],
  "xhrRequests": Array [],
}
`;

exports[`transmission/batched with sendBeacon API without batching must transmit larger payloads 1`] = `
Object {
  "beaconRequests": Array [
    Object {
      "data": "page	cart
id	42
m_featureFlags	{\"developerMode\":false}
longKey	aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
      "url": "http://eum.example.com",
    },
  ],
  "imageRequests": Array [],
  "xhrRequests": Array [],
}
`;

exports[`transmission/batched with sendBeacon API without batching must transmit multiple beacons 1`] = `
Object {
  "beaconRequests": Array [
    Object {
      "data": "ty	pl
page	landing
ts	123456789
error	false",
      "url": "http://eum.example.com",
    },
    Object {
      "data": "ty	xhr
page	cart
ts	123456789
error	true",
      "url": "http://eum.example.com",
    },
  ],
  "imageRequests": Array [],
  "xhrRequests": Array [],
}
`;

exports[`transmission/batched with unsupported sendBeacon API with batching must batch when window is visible 1`] = `
Object {
  "beaconRequests": Array [],
  "imageRequests": Array [],
  "xhrRequests": Array [
    Object {
      "open": Array [
        "POST",
        "http://eum.example.com",
        true,
      ],
      "requestHeader": Array [
        Array [
          "Content-type",
          "text/plain;charset=UTF-8",
        ],
      ],
      "responseType": "text",
      "send": Array [
        "ty	pl
page	landing
ts	123456789
error	false

ty	xhr
page	cart
ts	123456789
error	true",
      ],
      "timeout": 5432,
    },
  ],
}
`;

exports[`transmission/batched with unsupported sendBeacon API with batching must not attempt batching when window is hidden 1`] = `
Object {
  "beaconRequests": Array [],
  "imageRequests": Array [],
  "xhrRequests": Array [
    Object {
      "open": Array [
        "POST",
        "http://eum.example.com",
        true,
      ],
      "requestHeader": Array [
        Array [
          "Content-type",
          "text/plain;charset=UTF-8",
        ],
      ],
      "responseType": "text",
      "send": Array [
        "ty	pl
page	landing
ts	123456789
error	false",
      ],
      "timeout": 5432,
    },
    Object {
      "open": Array [
        "POST",
        "http://eum.example.com",
        true,
      ],
      "requestHeader": Array [
        Array [
          "Content-type",
          "text/plain;charset=UTF-8",
        ],
      ],
      "responseType": "text",
      "send": Array [
        "ty	xhr
page	cart
ts	123456789
error	true",
      ],
      "timeout": 5432,
    },
  ],
}
`;

exports[`transmission/batched with unsupported sendBeacon API with batching must send early when window becomes hidden 1`] = `
Object {
  "beaconRequests": Array [],
  "imageRequests": Array [],
  "xhrRequests": Array [],
}
`;

exports[`transmission/batched with unsupported sendBeacon API with batching must send early when window becomes hidden 2`] = `
Object {
  "beaconRequests": Array [],
  "imageRequests": Array [],
  "xhrRequests": Array [],
}
`;

exports[`transmission/batched with unsupported sendBeacon API without batching must do nothing when there is nothing to transmit 1`] = `
Object {
  "beaconRequests": Array [],
  "imageRequests": Array [],
  "xhrRequests": Array [],
}
`;

exports[`transmission/batched with unsupported sendBeacon API without batching must only transmit own props 1`] = `
Object {
  "beaconRequests": Array [],
  "imageRequests": Array [],
  "xhrRequests": Array [
    Object {
      "open": Array [
        "POST",
        "http://eum.example.com",
        true,
      ],
      "requestHeader": Array [
        Array [
          "Content-type",
          "text/plain;charset=UTF-8",
        ],
      ],
      "responseType": "text",
      "send": Array [
        "page	cart",
      ],
      "timeout": 5432,
    },
  ],
}
`;

exports[`transmission/batched with unsupported sendBeacon API without batching must transmit larger payloads 1`] = `
Object {
  "beaconRequests": Array [],
  "imageRequests": Array [],
  "xhrRequests": Array [
    Object {
      "open": Array [
        "POST",
        "http://eum.example.com",
        true,
      ],
      "requestHeader": Array [
        Array [
          "Content-type",
          "text/plain;charset=UTF-8",
        ],
      ],
      "responseType": "text",
      "send": Array [
        "page	cart
id	42
m_featureFlags	{\"developerMode\":false}
longKey	aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
      ],
      "timeout": 5432,
    },
  ],
}
`;

exports[`transmission/batched with unsupported sendBeacon API without batching must transmit multiple beacons 1`] = `
Object {
  "beaconRequests": Array [],
  "imageRequests": Array [],
  "xhrRequests": Array [
    Object {
      "open": Array [
        "POST",
        "http://eum.example.com",
        true,
      ],
      "requestHeader": Array [
        Array [
          "Content-type",
          "text/plain;charset=UTF-8",
        ],
      ],
      "responseType": "text",
      "send": Array [
        "ty	pl
page	landing
ts	123456789
error	false",
      ],
      "timeout": 5432,
    },
    Object {
      "open": Array [
        "POST",
        "http://eum.example.com",
        true,
      ],
      "requestHeader": Array [
        Array [
          "Content-type",
          "text/plain;charset=UTF-8",
        ],
      ],
      "responseType": "text",
      "send": Array [
        "ty	xhr
page	cart
ts	123456789
error	true",
      ],
      "timeout": 5432,
    },
  ],
}
`;
